experiments/train.py

```diff
@@ -0,0 +1,166 @@
+""" Generic training script used by all experiments.
+
+Added functionality must always be backwards compatible.
+
+Typical use:
+    # where "my_experiment_name" is the name of the folder in `experiment/`
+    python3 train.py my_experiment_name [HYDRA CLI ARGS]
+
+    ################################
+    # Resuming and Loading Weights #
+    ################################
+    # Common trainer overrides are at the root level.
+    # Only one (or None!) of these resume/finetune configurations should be set.
+    #
+    # If a provided path is relative, it MUST be relative to the specific 
+    # experiment's outputs/ directory.
+    #
+    # Example:
+    #      # structure:
+    #      #    pugh_torch/experiments/my_experiment/outputs/2020-09-27/10-43-40/default/version_0/checkpoints
+    #
+    #      python3 train.py my_experiment resume_checkpoint=2020-09-27/10-43-40/default/version_0/checkpoints 
+
+    # Load the latest checkpoint from the latest run.
+    # Note: this continues that training in a new output directory.
+    python3 train.py my_experiment_name +resume_last_run=true
+
+    # Load a specific checkpoint.
+    # This checkpoint path is relative
+    python3 train.py my_experiment_name +resume_checkpoint=path/to/checkpoint.ckpt
+
+    # Load only the weights of a checkpoint.
+    # Not strict about all the weights fitting the new model architecture.
+    python3 train.py me_experiment_name +fine_tune=path/to/checkpoint.ckpt
+"""
+
+import os
+import sys
+import logging
+from pathlib import Path
+
+import pugh_torch as pt
+from pugh_torch.helpers import working_dir, most_recent_checkpoint
+from pugh_torch.utils import TensorBoardLogger
+import pytorch_lightning as pl
+from pytorch_lightning import Trainer
+
+from omegaconf import OmegaConf
+import hydra
+
+log = logging.getLogger(__name__)
+
+_usage = """Pugh Torch Trainer
+
+Before running, copy the `template/` folder to get your experiment going.
+
+Usage:
+    train.py <experiment_name> [hydra-overrides]
+"""
+
+this_file_path = Path(__file__).resolve()
+this_file_dir = this_file_path.parent
+
+# Hack argv so that we have stronger control over hydra invocation
+# We use the first positional argument to dictate the experiment
+# folder
+try:
+    experiment_name = sys.argv.pop(1)
+except IndexError:
+    print(_usage)
+    sys.exit(1)
+
+experiment_path = this_file_dir / experiment_name
+if not experiment_path.is_dir():
+    raise FileNotFoundError(f"No such directory {experiment_path}")
+
+sys.path.insert(0, str(experiment_path))
+
+# We switch the working directory so that hydra's automatic path handling
+# happens within the experiment directory.
+with working_dir(experiment_path):
```

Comment:
you can just customize the Hydra working directory directly.
see [this](https://hydra.cc/docs/configure_hydra/workdir).

Reply:
now correct me if i'm wrong, but that would require me changing the configuration in each experiment folder. Here I am sharing a single unified `train.py` driver to handle all my hydra and pytorch setup.

Reply:
Can you point to something specific to understand what you mean?

Generally speaking, you can (and should) reuse your configuration and compose what you need.
specifically, you can include the same hydra config snippet from all your experiments.

base_config.yaml
```
hydra:
  run:
    dir: ...
```

experiment.yaml
```yaml
defaults:
 - base_config
 - other: defaults

other: stuff
```

Reply:
alright, i get what you're saying, but for my use case thats actually what i want to do.

1. Each experiments code is unique, the template is just some boilerplate/example code to get you going.
2. I want each experiment to be bundled with its default configuration so that each experiment is mostly isolated. This way the only backwards compatible elements between experiments are located in `train.py` and my package `pugh_torch`
3. I get that Hydra's way would be to shove all of that into a single config folder at `experiment/config`, and that would be convenient for some common configurations, but now each experiment's code/config is located in two places and has less-straightforward configuration (from a perspective of someone who is new to Hydra).

Simply put, I believe the current setup is more robust to people that don't fully understand Hydra from messing up other experiments than the one they are working on. It loses the benefits you mention, but it makes code reviews easier and getting new people started easier.